### PR TITLE
ci: add test visibility instrumentation to GitLab with pytest plugin

### PR DESCRIPTION
**Pull request by romainkomorndatadog**
This makes changes to enable Test Visibility in GitLab using `--ddtrace`.

It relies on GitLab having an agent at `${HOST_IP}:8196` and introduces the undocumented and "private" `_CI_DD_AGENT_URL` environment variable to make the "outer" `CIVisibility` tracer submit data via the agent.

In the past, the tracer ran in agentless mode which had more or less the same effect, but did not affect snapshot data.

`riot` environments for some tests do not pass with telemetry enabled (the fact that it was previously not enabled in CircleCI was actually an accident), so the `DD_INSTRUMENTATION_TELEMETRY_ENABLED` is set to a falsy value in this change (but is enabled for some `tracer` tests where it is needed).

Using the GitLab host allows us to run without an API key in our environment which is arguably worth it.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: .circleci/config.templ.yml
Comment (gnufede) [inline .circleci/config.templ.yml:189]: If I'm not mistaken, yaml is string by default, no?
```suggestion
                  DD_CIVISIBILITY_AGENTLESS_ENABLED: true
```

File: .circleci/config.templ.yml
Comment (gnufede) [inline .circleci/config.templ.yml:226]: Ditto

**Review by gnufede — APPROVED** [review]
LGTM

File: .circleci/config.templ.yml
Comment (romainkomorndatadog) [reply .circleci/config.templ.yml:189]: It is but I like my mistakes sprinkled throughout my PRs.

File: .gitlab/tests.yml
Comment (romainkomorndatadog) [inline .gitlab/tests.yml:30]:
```suggestion
```

File: .gitlab-ci.yml
Comment (romainkomorndatadog) [inline .gitlab-ci.yml:21]:
```suggestion
```

File: .gitlab/tests.yml
Comment (romainkomorndatadog) [inline .gitlab/tests.yml:11]:
```suggestion
```

**Review by brettlangdon — APPROVED** [review]
changes to files owned by core team lgtm

I did not thoroughly review the changes to Test Visibility product, I am unsure the impact these changes will have on customers.

**Review by emmettbutler — APPROVED** [review]

File: riotfile.py
Comment (erikayasuda) [inline riotfile.py:106]: It looks like `DD_CIVISIBILITY_ITR_ENABLED` defaults to True so that was removed. 

But for `DD_CIVISIBILITY_AGENTLESS_ENABLED` and `DD_CIVISIBILITY_CODE_COVERAGE_ENABLED`, I just see these set to `1` in the `hatch.toml` file but not anywhere else. Are they unused?

File: .gitlab/tests.yml
Comment (erikayasuda) [inline .gitlab/tests.yml:12]: Do we set this for hatch but not for riot because it's already set in [riotfile.py](https://github.com/DataDog/dd-trace-py/blob/main/riotfile.py#L101)?

**Review by ZStriker19 — APPROVED** [review]

File: .gitlab-ci.yml
Comment (erikayasuda) [inline .gitlab-ci.yml:23]: To clarify: the reason we need to add this new `_CI_DD_AGENT_URL` var is because with GitLab, we (currently) don't have a way to spin up a separate service/test agent, but we did with CircleCI?

**Review by erikayasuda — APPROVED** [review]
Sorry just a bunch of clarifying questions, otherwise LGTM if we can get some of the (I think) flaky tests to pass

File: .gitlab/tests.yml
Comment (romainkomorndatadog) [reply .gitlab/tests.yml:12]: Correct.

File: riotfile.py
Comment (romainkomorndatadog) [reply riotfile.py:106]: `DD_CIVISIBILITY_AGENTLESS_ENABLED` is being removed because we don't want it in the `riotfile.py` (and we want it to be falsy anyway).

`DD_CIVISIBILITY_CODE_COVERAGE_ENABLED` is actually no longer a valid env var at all.

File: .github/workflows/testrunner.yml
Comment (romainkomorndatadog) [reply .github/workflows/testrunner.yml:19]: It was not.

File: docker/Dockerfile
Comment (romainkomorndatadog) [reply docker/Dockerfile:61]: I found https://www.reddit.com/r/mariadb/comments/1e5f38t/mariadb_repo_setup_is_broken_leads_to_unavailable/ and decided it's a known issue and, at best, we don't want it to break other potential `apt` usage, so I stuck it at the end.

File: .gitlab-ci.yml
Comment (romainkomorndatadog) [reply .gitlab-ci.yml:23]: In CircleCI, we were using agentless mode, so we didn't care what the agent URL was, but I decided it would be better to use an agent in GitLab since a- it's there, and b- it means we don't have to get an API key.

The `CIVisibility` tracer needs "its own" agent URL because otherwise it'll try to send traces to the test agent, in the snapshot test jobs (and the existing use of `_DD_CIVISIBILITY_USE_CI_CONTEXT_PROVIDER` already forces it to create a different tracer).

File: ddtrace/internal/ci_visibility/writer.py
Comment (romainkomorndatadog) [reply ddtrace/internal/ci_visibility/writer.py:110]: I like it this way because it makes it clear that we want to use the same URL as the `intake_url` when we have an agent, or when we're hardcoding the agentless URL.

**Pull request merged**
